swupd/rename.go

```diff
@@ -0,0 +1,183 @@
+	for ax, rx := 0, 0; ax < len(pa) && rx < len(pr); {
+		af := pa[ax]
+		rf := pr[rx]
+		switch {
+		case af.n < rf.n:
+			ax++
+		case af.n > rf.n:
+			rx++
+		default: // Equal truncated name
+			linkrenamepair(af.f, rf.f)
+			ax++
+			rx++
+		}
+	}
+}
+
+// linkrenamepair links two files together
+func linkrenamepair(a, b *File) {
+	a.DeltaPeer = b
+	b.DeltaPeer = a
+	a.Rename = true
+	b.Rename = true
+}
+
+// trimrenamed returns an slice which has had files with DeltaPeers purged
+func trimrenamed(a []*File) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		if f.Status == StatusGhosted {
+			panic("Found ghosted file " + f.Name)
+		}
+		if f.DeltaPeer == nil {
+			r = append(r, f)
+		}
+	}
+	return r
+}
+
+// trimsmall returns an slice which has had files with small files removed
+func trimsmall(a []*File, minsize int64) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		if f.Info == nil {
+			panic("Missing f.Info for " + f.Name)
+		}
+		if f.Info.Size() > minsize {
+			r = append(r, f)
+		}
+	}
+	return r
+}
+
+// pairednames holds a *File and another "name", which is used for various
+// alterations of the filename for matching purposes. In particular removing
+// digits so a new release of a shared library is matched to previous ones
+type pairednames struct {
+	f *File
+	n string // Filename with digits removed
+}
+
+// stripdigits is used with strings.Map to remove digits from filename
+func stripdigits(r rune) rune {
+	if r >= '0' && r <= '9' {
+		return -1 // strings.Map removes negative values
+	}
+	return r
+}
+
+// makepairs returns a sorted by name, disregarding digits
+// secondary sort key is the original name. This is mainly inteneded for
+// cases like with python we have the same filename under both python3.6 and python2.7
+// it would look weird to rename the old python2.7/file1 to python3.6/file2 and at
+// the same time rename the old python3.6/file1 to python2.7/file2.
+func makepairs(a []*File) []pairednames {
+	pair := make([]pairednames, len(a))
+	for i, f := range a {
+		pair[i].f = f
+		pair[i].n = strings.Map(stripdigits, f.Name)
+	}
+	sort.Slice(pair, func(i, j int) bool {
+		if pair[1].n == pair[j].n { // Same striped name, sort on original name
+			// mainly for python2.x vs 3.x
+			return pair[i].f.Name < pair[j].f.Name
+		}
+		return pair[i].n < pair[j].n
+	})
+	return pair
+}
+
+// fixupstatfields adds the missing stat fields
+// construct the path to the old chroot by Joining the c.imageBase
+// file.Previous, bundle name, and file.Name fields
+func fixupstatfields(needed []*File, m *Manifest, c *config) error {
+	var bundleChroot string
+	for i := range needed {
+		if needed[i].Info != nil {
+			continue
+		}
+		if bundleChroot == "" {
+			bundleChroot = filepath.Join(c.imageBase, fmt.Sprint(m.Header.Previous), m.Name)
+		}
+		path := filepath.Join(bundleChroot, needed[i].Name)
+		fi, err := os.Stat(path)
+		if err != nil {
+			panic(err)
+			return err
```

Comment:
you `panic` and `return` here, then `panic` in the caller as well.

Reply:
You can tell I was getting frustrated with the lack of Info! This panic removed. I need to remove the other one soon.